### DISCUSSION TITLE
How to override global dependencies on a single endpoint?

### DISCUSSION THREAD

**Discussion by PatrikTrelsmo-izettle** [fastapi/fastapi#7089] [Questions]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case, copied from https://fastapi.tiangolo.com/tutorial/dependencies/global-dependencies/:

```Python
from fastapi import Depends, FastAPI, Header, HTTPException


async def verify_token(x_token: str = Header(...)):
    if x_token != "fake-super-secret-token":
        raise HTTPException(status_code=400, detail="X-Token header invalid")


async def verify_key(x_key: str = Header(...)):
    if x_key != "fake-super-secret-key":
        raise HTTPException(status_code=400, detail="X-Key header invalid")
    return x_key


app = FastAPI(dependencies=[Depends(verify_token), Depends(verify_key)])


@app.get("/items/")
async def read_items():
    return [{"item": "Portal Gun"}, {"item": "Plumbus"}]


@app.get("/users/")
async def read_users():
    return [{"username": "Rick"}, {"username": "Morty"}]

# This function shouldn't be authenticated because it is a health check endpoint
@app.get("/healthcheck")
async def healthcheck():
    return "ok"
```

### Description

Is there a way to "skip" or override global dependencies on a single endpoint? I really like the new global dependencies since it makes it easy to enforce authentication by design rather than remembering to add it to every single endpoint. 

However, in many deployment scenarios we need to expose a "healthcheck" endpoint, which shouldn't be authenticated. I've looked but not found a way of doing so in FastAPI. I could place my app behind Nginx and place my health check there, but it doesn't feel quite right. 

### Environment

* OS: [e.g. Linux / Windows / macOS]: Any (unix-based Docker container)
* FastAPI Version [e.g. 0.3.0]: 0.62.0

* Python version:

Any, I presume. Using 3.7 personally. 

### Additional context
-

**Comment by falkben** [comment]
Can you use a separate router for the healthcheck, and put everything else on another router? Then put your "global" dependencies on the second router?

**Comment by PatrikTrelsmo-izettle** [comment]
That sounds like a very good idea @falkben! Thanks for the suggestion. I'll try it out and update with the results 👍

**Comment by PatrikTrelsmo-izettle** [comment]
It worked just fine. It's not entirely as safe (a developer might accidentally use `@app.get` instead of `@router.get` and thereby bypass the authentication), but it does solve my use-case well enough for this ticket to be closed. Thanks again!

**Comment by tiangolo** [comment]
Thanks for the help here @falkben ! :clap: :bow: 

Thanks for reporting back and closing the issue @PatrikTrelsmo-izettle  :+1:

**Comment by tearf001** [comment]
APIRouter has an initalize-option `route_class`, which not introduced by `fastapi office docs`.
You can override the class. here comes the [source codes](https://github.com/tiangolo/fastapi/blob/master/fastapi/routing.py#LINE274):
```py
class APIRoute(routing.Route):
    def __init__(
        self,
        path: str,
        endpoint: Callable,
        *,
        response_model: Type[Any] = None,
        status_code: int = 200,
        tags: List[str] = None,
        dependencies: Sequence[params.Depends] = None, # here todo 
        ...more options
    )
```

**Comment by tearf001** [comment]
> APIRouter has an initalize-option `route_class`, which not introduced by `fastapi office docs`.
> You can override the class. here comes the [source codes](https://github.com/tiangolo/fastapi/blob/master/fastapi/routing.py#LINE274):
> 
> ```python
> class APIRoute(routing.Route):
>     def __init__(
>         self,
>         path: str,
>         endpoint: Callable,
>         *,
>         response_model: Type[Any] = None,
>         status_code: int = 200,
>         tags: List[str] = None,
>         dependencies: Sequence[params.Depends] = None, # here todo 
>         ...more options
>     )
> ```

```py
class AuthenticatedRoute(APIRoute):
    def __init__(self, *args, **kwargs):
        dependencies = list(kwargs.pop('dependencies', None) or [])
        dependencies.insert(0, Depends(has_admin_role))
        kwargs['dependencies'] = dependencies
        super(AuthenticatedRoute, self).__init__(*args, **kwargs)

router = APIRouter(route_class=AuthenticatedRoute)
@router.get("/api/some-resource", dependencies=[Depends(some_more_other_resource)])```
def view():
   pass

**Comment by tearf001** [comment]
There should be a better solve to place a `router-scope-global`  dependencies. 
for example:
```py
router = APIRouter(scope_dependencies=[Depends(...),..])
```

**Comment by bhaktatejas922** [comment]
> It worked just fine. It's not entirely as safe (a developer might accidentally use `@app.get` instead of `@router.get` and thereby bypass the authentication), but it does solve my use-case well enough for this ticket to be closed. Thanks again!

How exactly did you use a second router here? If you could post a simple example thatd be great

**Comment by kristjanvalur** [comment]
I've been wondering about this too.  Automatic dependencies can now be added both globally and to individual routers.
However, it would be nice to be able to disable them, on individual endpoints, maybe something like

```python
@router.disable_dependencies(global_depend).get("/api/resource")
def foo(): pass

# or
@router.get("/api/resource", disabled_dependencies=[Depends(global_depend)])
def foo(): pass
```

This avoids accumulating a system of special routers with different default dependencies on them.

To avoid parameter permutation explosion, it could be cleaner to add an argument to `Depends`:

```python
@router.get("/api/resource", dependencies=[Depends(global_depend, disable=True)])
def foo(): pass
```

The final suggestion is probably the simplest one to implement, I'll try to see if I can make it work.

**Reply by kristjanvalur** [reply]
See #5433 for a suggested implementation

**Comment by kristjanvalur** [comment]
Another suggested solution is in pr #5667, for combining intherited non-param security objects.

**Comment by jcphlux** [comment]
Just adding my 2 cents here as my search for a solution got me here first.
This might not be the best solution but it is working for me.
I have an endpoint folder with an __init__.py that looks like this
```python
from fastapi import APIRouter, Depends

from app.api.dependencies.auth import get_current_active_user

from .endpoints.auth import router as auth_router, unsecure_router as unsecure_auth_router
from .endpoints.admin import router as admin_router

api_secure_router = APIRouter(dependencies=[Depends(get_current_active_user)])

# add secure routes here
api_secure_router.include_router(auth_router)
api_secure_router.include_router(admin_router)

api_router = APIRouter()
api_router.include_router(api_secure_router)

# add unsecure routes here
api_router.include_router(unsecure_auth_router)
```

and then in my app I add my endpoints
```python
from fastapi import FastAPI

from .api import api_router

app = FastAPI()
app.include_router(api_router)
```

**Comment by slingshotsys** [comment]
+1 on this.  A per-endpoint dependency override or disable setting would be super useful for security scenarios where all routes in a module need to be secure except for one.

**Comment by astro-angelfish** [comment]
for developers that seeks for a temporary workaround, i've created a parallel route to accept unauthenticated requests

```
router = APIRouter(
    prefix="/api",
    dependencies=[Depends(check_auth_or_sth)],
)

router_unsafe = APIRouter(
    prefix="/api",
)
```

put authenticated requests into `router` and others to `router_unsafe` should work, however you have to include both router.

i cannot say this is the best or a beautiful solution, but at least it works
